File: Projects/Project_06_2D_Wave/For_Students/matlab/create_rhs.m
%----------------------------------------------------------------------%
%This subroutine builds the RHS vector for the Weak Form CG/DG Methods
%on Quadrilateral Elements for the 2D Advection Equation.
%Written by Francis X. Giraldo on 5/2021
%           Naval Postgraduate School
%           Department of Applied Mathematics
%           Monterey, CA 93943-5502
%----------------------------------------------------------------------%
function rhs = create_rhs(q,u,v,ksi_x,ksi_y,eta_x,eta_y,jac,...
		 wnq,psi,dpsi,intma,iperiodic,Mmatrix,face,normals,jac_face,...
         mapL,mapR,npoin,nelem,nface,ngl,space_method,flux_method,form_method)

%------------Students Add Your Routines Here---------------%
%------------Students Add Your Routines Here---------------%
%Compute Volume Integral: Differentiation matrix contribution
rhs = create_rhs_volume(q,u,v,ksi_x,ksi_y,eta_x,eta_y,jac,...
    wnq,psi,dpsi,intma,iperiodic,npoin,nelem,ngl,form_method);
%Construct Communicator: DSS for CG / Fluxes for DG
if strcmp(space_method,'dg')
    rhs = create_rhs_flux(rhs,q,u,v,face,normals,jac_face, ...
          wnq,nface,ngl,mapL,mapR,intma,flux_method,form_method);
end %if
%------------Students Add Your Routines Here---------------%
%------------Students Add Your Routines Here---------------%

%Multiply by Inverse Mass
rhs = rhs./Mmatrix;

%Periodicity
for i=1:npoin
    rhs(i)=rhs(iperiodic(i));
end

File: Projects/Project_06_2D_Wave/For_Instructors/julia/julia_Local_Matrices/create_rhs.jl
#=
---------------------------------------------------------------------
This function computes the RHS vector for a unified CGDG method for the 
2D Wave Equation.

Written by F.X. Giraldo on July 12, 2021
           Department of Applied Mathematics
           Naval Postgraduate School
           Monterey; CA 93943-5216
---------------------------------------------------------------------
=#

include("create_rhs_volume.jl")
include("create_rhs_flux.jl")

function create_rhs(q,u,M,De_x,De_y,intma,periodicity,face,mapL,mapR,normals,jac_face,ωq,space_method,DFloat)
 
    #---------------------------Students Add their Contributions Here----------------------------------------
    #---------------------------Students Add their Contributions Here----------------------------------------
    #Construct Volume integral contribution
    rhs = create_rhs_volume(q,u,De_x,De_y,intma,periodicity,DFloat)

    #Construct Flux integral contribution (will also work for CG but cancels completely due to periodicity)
    if (space_method == "dg")
        rhs = create_rhs_flux!(rhs,q,u,face,normals,jac_face,ωq,mapL,mapR,intma,periodicity,DFloat)
    end
    #---------------------------Students Add their Contributions Here----------------------------------------
    #---------------------------Students Add their Contributions Here----------------------------------------

    #Multiply by inverse Mass matrix
    rhs=M\rhs

    #Apply Periodicity
    Npoin=length(q)
    for i=1:Npoin
        rhs[i]=rhs[periodicity[i]]
    end
    
    #Return arrays
    return(rhs)
end
